File: 225-implement-stack-using-queues/225-implement-stack-using-queues.py
class MyStack:

    def __init__(self):
        self.q1 = deque()

    def push(self, x: int) -> None:
        self.q1.append(x)
        
    def pop(self) -> int:
        for i in range(len(self.q1) - 1):
            self.q1.append(self.q1.popleft())
        return self.q1.popleft()

    def top(self) -> int:
        for i in range(len(self.q1) - 1):
            self.q1.append(self.q1.popleft())
        res = self.q1[0]
        self.q1.append(self.q1.popleft())
        return res

    def empty(self) -> bool:
        return not self.q1


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: 225-implement-stack-using-queues/225-implement-stack-using-queues.java
class MyStack {
    Queue<Integer> q1, q2;
    public MyStack() {
        q1 = new LinkedList<>();
        q2 = new LinkedList<>();
    }
    
    public void push(int x) {
        q1.offer(x);
    }
    
    public int pop() {
        while(q1.size() > 1) {
            q2.offer(q1.poll());
        }
        int res = q1.poll();
        Queue<Integer> temp_q = q1;
        q1 = q2;
        q2 = temp_q;
        return res;
    }
    
    public int top() {
        while(q1.size() > 1) {
            q2.offer(q1.poll());
        }
        int res = q1.poll();
        q2.offer(res);
        Queue<Integer> temp_q = q1;
        q1 = q2;
        q2 = temp_q;
        return res;
    }
    
    public boolean empty() {
        return q1.isEmpty() && q2.isEmpty();
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * boolean param_4 = obj.empty();
 */